Proofreading MO Guide (#11605)
openvinotoolkit/openvino: docs/MO_DG/prepare_model/customize_model_optimizer/Customize_Model_Optimizer.md

<a name="model-optimizer-extensibility"></a>Model Optimizer extensibility mechanism enables support of new operations and custom transformations to generate the optimized intermediate representation (IR) as described in the
[Deep Learning Network Intermediate Representation and Operation Sets in OpenVINO™](../../IR_and_opsets.md). This

> mechanism is a core part of Model Optimizer. Model Optimizer itself uses this huge set of examples of adding custom logic to support your model.

Author: It does :) As for me, this "bird" can fly away too :)

> mechanism is a core part of Model Optimizer. Model Optimizer itself uses this huge set of examples of adding custom logic to support your model.

Reviewer: ```suggestion
mechanism is a core part of Model Optimizer, which is used internally as a huge set of examples showing how to add custom logic to support your model.
```

"Under the hood" means "in its internal working" so maybe we should leave that part that as well, but phrased more clearly :)

![Graph After Attributes Extraction](../../../img/MO_graph_after_extractors.png)

The only difference in the graph from the previous step is that nodes contain dictionary with extracted attributes and
operation-specific attributes needed for Model Optimizer. However, starting from this step, Model Optimizer does not
need the original representation of the operation/model and uses just Model Optimizer representation (there are some

> very special cases when Model Optimizer still uses the *`pb`* attribute and they are partially covered in this
> article). Detailed list of common node attributes and their values is provided below in the
> [Model Optimizer Operation](#extension-operation) section.

Reviewer 2: ```suggestion
peculiar cases in which Model Optimizer still uses the *`pb`* attribute, covered in this
article partially). A detailed list of common node attributes and their values is provided below in the
[Model Optimizer Operation](#extension-operation) section.
```

This layout change is disabled automatically if the model does not have operations that OpenVINO&trade needs to execute in the NCHW layout, for example, Convolutions in NHWC layout.

It is still possible to force Model Optimizer to do a layout change, using *`--disable_nhwc_to_nchw`* command-line parameter, although it is not advised.

> The layout change is a complex problem, which is not explained in detail in this article. A very brief

Reviewer 2: ```suggestion
Layout change is a complex problem and will be addressed here very briefly. For more details on how it works, refer to the source code of the transformations mentioned in the below summary of the process: 
```

> explanation of this process is provided below:

Reviewer 2: ```suggestion
```

1. Model Optimizer changes output shapes of most of operations producing 4D and 5D (four dimensional and five
dimensional) tensors as if they were in NHWC layout to NCHW layout: *`nchw_shape = np.array(nhwc_shape)[0, 3, 1, 2]`* for
4D and *`nchw_shape = np.array(nhwc_shape)[0, 4, 1, 2, 3]`* for 5D. This permutation does not happen for some operations
with specific conditions identified during a model conversion.
2. Model Optimizer inserts [Gather](../../../ops/movement/Gather_1.md) operations to the sub-graph relates to shapes
calculation in order to perform shape calculation in a correct layout.
3. Model Optimizer inserts [Transpose](../../../ops/movement/Transpose_1.md) operations for some operations with
specific conditions, identified during a model conversion, to produce correct inference results.

The list of main transformations responsible for a layout change are: *`extensions/middle/ApplyPermutations.py`*,
*`extensions/middle/InsertLayoutPropagationTransposes.py`*, *`extensions/middle/MarkSubgraphsWithCorrectLayout.py`*,
*`extensions/middle/ApplyNHWCtoNCHWpermutation.py`* and *`extensions/middle/LayoutChangeForConstantShapePaths.py`*.

> For more details on how the layout change works, refer to the source code of these transformations.

Reviewer 2: ```suggestion
```

class ConstExtractor(FrontExtractorOp):
    # The "op" class attribute defines a type of the operation in the framework (in this case it is a TensorFlow), 
    # for which the extractor should be triggered.
    op = 'Const'
    enabled = True  # The flag that indicates that this extractor is enabled.

    @classmethod
    def extract(cls, node):  # The entry point of the extractor.

>         # The node.pb attribute stores the TensorFlow representation of the operation, which is a Protobuf message of the

Author: ```suggestion
        # The *`node.pb`* attribute stores the TensorFlow representation of the operation, which is a Protobuf message of the
```

            op_params.update(replacement_descriptions)
            region_layer = RegionYoloOp(graph, op_params)
            region_layer_node = region_layer.create_node([last_node])

>             # Here you remove 'axis' from 'dim_attrs' to avoid permutation from axis = 1 to axis = 2.

Author: ```suggestion
            # In here, 'axis' from 'dim_attrs' can be removed to avoid permutation from axis = 1 to axis = 2.
```
